src/app/bmc_application.rs

```diff
@@ -121,168 +149,104 @@ impl BmcApplication {
     }
 
     async fn initialize_power(&self) -> anyhow::Result<()> {
-        // power on nodes
-        if let Ok(enabled_nodes) = self.app_db.get::<u8>(ACTIVATED_NODES_KEY).await {
-            self.power_node(enabled_nodes, 0b1111).await
-        } else {
+        if self.app_db.get::<u8>(ACTIVATED_NODES_KEY).await.is_err() {
             // default, given a new app persistency
-            self.app_db.set::<u8>(ACTIVATED_NODES_KEY, 0).await
+            self.app_db.set::<u8>(ACTIVATED_NODES_KEY, 0).await?;
         }
+        self.power_on().await
     }
 
-    async fn initialize_usb_mode(&self) -> std::io::Result<()> {
-        let node = self
+    async fn initialize_usb_mode(&self) -> anyhow::Result<()> {
+        let config = self
             .app_db
-            .get::<NodeId>(USB_NODE_KEY)
+            .get::<UsbConfig>(USB_CONFIG)
             .await
-            .unwrap_or(NodeId::Node1);
-        let res = self.pin_controller.select_usb(node);
-
-        let route = self
-            .app_db
-            .get::<UsbRoute>(USB_ROUTE_KEY)
-            .await
-            .unwrap_or(UsbRoute::UsbA);
-        let res2 = self.pin_controller.set_usb_route(route);
-
-        let mode = self.app_db.get::<u8>(USB_MODE_KEY).await.unwrap_or(0b1111);
-        let res3 = self.pin_controller.inner_set_usb_mode(mode);
-
-        res.and(res2).and(res3)
+            .unwrap_or(UsbConfig::UsbA(NodeId::Node1, false));
+        self.configure_usb(config).await.context("usb configure")
     }
 
-    /// Helper function that returns the new state of ATX power
-    fn need_atx_change(current_node_state: u8, next_node_state: u8) -> Option<bool> {
-        if current_node_state == 0 && next_node_state > 0 {
-            // power down
-            Some(true)
-        } else if current_node_state > 0 && next_node_state == 0 {
-            // power up
-            Some(false)
+    /// routine to support legacy API
+    pub async fn get_node_power(&self, node: NodeId) -> anyhow::Result<bool> {
+        let state = self.app_db.get::<u8>(ACTIVATED_NODES_KEY).await?;
+        if self.nodes_on.load(Ordering::Relaxed) {
+            Ok(state & node.to_bitfield() != 0)
         } else {
-            // dont do anything
-            None
+            Ok(false)
         }
     }
 
-    pub async fn get_node_power(&self, node: NodeId) -> anyhow::Result<bool> {
-        let state = self.power_state.lock().await;
-        Ok(*state & node.to_bitfield() != 0)
-    }
-
     /// This function is used to active a given node. Call this function if a
     /// module is inserted at that slot. Failing to call this method means that
     /// this slot is not considered for power up and power down commands.
-    pub async fn activate_slot(&self, node: NodeId, on: bool) -> anyhow::Result<()> {
-        ensure!(node.to_bitfield() != 0);
+    pub async fn activate_slot(&self, node_states: u8, mask: u8) -> anyhow::Result<()> {
```

Comment:
This interface still grinds a bit. But this comes from the fact that we use this routine to do two things:
* Signal that there is a module inserted at the given slot. (`ACTIVATED_NODES`).  This information we use later on a power on/off or power toggle so that we know which particular nodes to power.
* power on/off the particular node.

Another option would be to drop the activation logic. Then this routine will be more straightforward power nodes on/off. It would mean that the KEY1 button always toggles all nodes.

Reply:
You can simplify these routines and/or provide a simpler method with the original intent. For example, when one calls `activate_slot()` or `toggle_power_states()` there are several lookups of `ACTIVATED_NODES_KEY`, while 1 should suffice.

Reply:
im not sure if make it more simple by squashing calls the calls to ACTIVATED_NODES_KEY. I tried to incorporate your comments in my latest push. Could you see if its more to your liking?